File: 994-rotting-oranges/994-rotting-oranges.rs
use std::collections::VecDeque;

impl Solution {
    pub fn oranges_rotting(mut grid: Vec<Vec<i32>>) -> i32 {
        let mut q = VecDeque::new();
        let (m, n) = (grid.len(), grid[0].len());
        
        let (empty, fresh, rotten) = (0, 1, 2);
        let mut fresh_oranges = 0;
        let mut mins = 0;
        
        for i in 0..m{
            for j in 0..n{
                if grid[i][j] == rotten{
                    q.push_back((i, j));
                }else if grid[i][j] == fresh{
                    fresh_oranges +=1;
                }
            }
        }
        
        
        while !q.is_empty(){
            let mut any_spoiled = false;
            
            for _ in 0..q.len(){
                let (i, j) = q.pop_front().unwrap();
                
                for (x, y) in [(i + 1, j), (i - 1, j), (i, j + 1), (i, j-1)]{
                    
                    if x!= usize::MAX && y!= usize::MAX && x < m && y < n && grid[x][y] == fresh{
                        q.push_back((x, y));
                        fresh_oranges -=1;
                        
                        grid[x][y] = rotten;
                        any_spoiled = true;
                    }
                }
            }
            
            if any_spoiled{
                mins +=1;
            }
        }
        
        if fresh_oranges == 0{
            mins
        }else{
            -1
        }
    }
}

File: 994-rotting-oranges/994-rotting-oranges.py
from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rotten = deque()
        fresh = 0
        (m, n) = (len(grid), len(grid[0]))
        
        for i in range(0, m):
            for j in range(0, n):
                if grid[i][j] == 2:
                    rotten.append((i, j))
                elif grid[i][j] == 1:
                    fresh+=1
                
        dirs = [(1, 0),(0, 1), (-1, 0), (0, -1)]
        days = 0
        while len(rotten) > 0:
            changed = False
            # print(rotten)
            for _ in range(0, len(rotten)):
                (x, y) = rotten.popleft()
                
                for (dx, dy) in dirs:
                    i = x + dx
                    j = y + dy
                    
                    if i >= 0 and j >= 0 and i < m and j < n:
                        
                        if grid[i][j] == 1:
                            grid[i][j] = 2
                            fresh-=1
                            changed = True
                            rotten.append((i, j))
                            
            if changed:
                days +=1
        
        if fresh > 0:
            return -1
        
        return days
                    
            